pdf-cdn.md

```diff
@@ -0,0 +1,222 @@
+`https://arxiv.org/pdf/2001.00021` to
+`https://arxiv.org/pdf/2001.00021.pdf` in order to have a reasonably
+named download. There are also redirects from HTTP to HTTPS.
+
+### Limitations to GCP LB rules
+The GCP LB has URL mapping and rules but
+they are limited. "Path matchers (and URL maps in general) do not
+offer features that function like Apache LocationMatch directives. If
+you have an application that generates dynamic URL paths that have a
+common prefix, such as /videos/hd-abcd and /videos/hd-pqrs, and you
+need to send requests made to those paths to different backend
+services, you might not be able to do that with a URL map. For simple
+cases containing only a few possible dynamic URLs, you might be able
+to create a path matcher with a limited set of path rules. For more
+complex cases, you need to do path-based regular expression matching
+on your backends." see
+[https://cloud.google.com/load-balancing/docs/url-map-concepts#wildcards-regx-dynamic](https://cloud.google.com/load-balancing/docs/url-map-concepts#wildcards-regx-dynamic)
+
+### No custom 404 handling in LB
+I also was looking to see if we could
+route all 404s to a service but didn't find anything like that. This issue indicates that this is not a feature:
+[Allow Cloud Storage users to specify an http status code for 404 pages](https://issuetracker.google.com/issues/143348697)
+
+### CDN Cache keys
+[https://cloud.google.com/cdn/docs/caching#cache-keys](https://cloud.google.com/cdn/docs/caching#cache-keys) GCP can
+include headers and cookies as part of the cache key. Some of the GS
+query parameters are set as default parts of the cache key when the
+back end is a storage bucket.
+
+### GS Buckets versions
+CDN can cache based on query parameter
+`?generation=1234`. It sounds like we can alter earlier versions. The
+generations can be listed.
+[https://cloud.google.com/storage/docs/object-versioning](https://cloud.google.com/storage/docs/object-versioning)
+
+### No symbolic links in GS
+GS does not support symbolic links. "...there are no plans for gsutil rsync to support operating system-specific file types like symlinks" [from the gsutil rsync docs](https://cloud.google.com/storage/docs/gsutil/commands/rsync)
+
+### Downloaded files end with .pdf
+We redirect to a URL that ends with `.pdf`. This is likely to make the
+downloaded file more useful since OSs will recognize it as a PDF and
+open it with the appropriate viewer. We could achieve the correct
+filename with a [Content-Disposition HTTP header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition).
+
+## Timing
+The normal path at arxiv.org to get a PDF uses a redirect. The abs
+pages have `arxiv.org/pdf/astro-ph/0606003` and that gets redirected to
+`arxiv.org/pdf/astro-ph/0606003.pdf`. 
+
+A request for `arxiv.org/pdf/astro-ph/0606003` is: 58ms for first
+redirect, 655ms for the 2nd request that does the download.
+
```

Comment:
Is the 655ms for the current complicated get_ps logic? or is this get_ps logic plus download?